README.md
# cre
Excel variant report generator and scripts to process WES data (cram/bam/fastq -> variant calls -> annotated variant calls -> prioritized variants -> excel report). Uses results from [bcbio variant2](https://bcbio-nextgen.readthedocs.io/en/latest/contents/pipelines.html#germline-variant-calling) germline variant calling pipeline.

# 1. Installation

1.1 Install **bcbio-nextgen**.
*  [Installation manual](https://bcbio-nextgen.readthedocs.io/en/latest/contents/installation.html)
*  [Installation and update examples](https://github.com/naumenko-sa/cre/blob/master/cre.bcbio.upgrade.sh)

Use HPC or server. Add bcbio to PATH and PYTHONPATH in .bash_profile:
```
export PATH=[installation_path]/tools/bcbio/bin:[installation_path]/tools/bcbio/anaconda/bin:$PATH
export PYTHONPATH=[installation_path]/tools/bcbio/anaconda/lib/python2.7:$PYTHONPATH
```
bcbio installs many other useful tools (including java and R) and datasets through bioconda and cloudbiolinux.
  
1.2 Clone **cre** to ~/cre and add it to PATH.
```
cd
git clone https://github.com/naumenko-sa/cre
```
.bash_profile: `export PATH=~/cre:$PATH`

1.3 (Optional) Install/update OMIM.

By default CRE uses [cre/data/omim.txt](../master/data/omim.txt) and [cre/data/omim.inheritance.csv](../master/data/omim.inheritance.csv).
* Goto https://omim.org/downloads/ and request the latest database.
* In a couple of days you will receive: genemap2.txt, genemap.txt, mim2gene.txt, mimTitles.percent.txt, mimTitles.txt, morbidmap.txt. 
* Preprocess OMIM with [cre.omim.sh](../master/cre.omim.sh): `cd OMIM_DIR;~/cre/cre.omim.sh`
  
Result - omim.txt with omim description of diseases related to ~ 4000 genes
We use the improved OMIM inheritance table from [https://www.cs.toronto.edu/~buske/cheo/](https://www.cs.toronto.edu/~buske/cheo/).Download the second file with inheritance mappings. It references genes by gene name (symbol) rather than by Ensembl_id which is a requirement for CRE. Most gene names (symbols) could be mapped automatically with Ensembl biomart [genes.R](https://github.com/naumenko-sa/bioscripts/blob/master/genes.R), but some genes (not many) might need manual curation to assign the correct ENSEMBL_ID.

1.4 (Optional) Install/update Orphanet: `cd ~/cre/data; ~/cre/cre.orphanet.sh`
Orphanet provides descriptions for ~3600 genes:. By default CRE uses [orphanet.txt](../master/data/orphanet.txt)
1.5 (Optional) Update Gnomad gene contraint scores: `Rscript ~/cre/cre.gnomad_scores.R`
By default using [~/cre/data/gnomad_scores.csv](../master/data/gnomad_scores.csv)

1.6 (Optional) Imprinted genes.
By default using [~/cre/data/imprinting.txt](../master/data/imprinting.txt).

1.7 (Optional) Install HGMD pro database
Install HGMD pro and dump information with [~/cre/cre.hgmd2csv.sql](../master/cre.hgmd2csv.sql).

# 2. (optional) Alignment to grch37 with decoy

By default, bcbio does not have decoy in grch37 reference, decoy is supported only in grch38. Using decoy improves FDR by ~0.5%. Two step approach could be applied to use decoy in bcbio https://github.com/bcbio/bcbio-nextgen/issues/2489:
- install custom grch37d5 reference with decoy:  [cre.bcbio.custom_genome.sh](../master/cre.bcbio.custom_genome.sh)
- run alignment step vs grch37d5 reference: `cre.prepare_bcbio_run.sh <project> align_decoy`
- keep bam file aligned vs grch37d5 for storage
- run variant calling with noalt_calling and bam_clean: remove_extracontigs (SV calling in WGS required additional processing of decoy aligned bam file, see crg).
- when saving project to the storage, use bam files aligned to decoy

# 3. Set up bcbio project for alignment, variant caling and annotation

* Prepare input files: family_sample_1.fq.gz, family_sample_2.fq.gz, or family_sample.bam and place them into family/input folder.
* There might be many samples in a family(project).
* TEST: NIST Ashkenazim trio: ftp://ftp-trace.ncbi.nlm.nih.gov/giab/ftp/data/AshkenazimTrio (download OsloUniversityHospital exomes).
* run [cre.prepare_bcbio_run.sh](../master/cre.prepare_bcbio_run.sh) [family].

By default it uses [bcbio.templates.wes.yaml](../master/bcbio.templates.wes.yaml) config with following features:
  * 4 callers, the order is important, because variant metrics in ensemble calling (like AD) are picked up from the first caller in the list (gatk)
  * ensemble calling
  * realignment and recalibration. There is no much gain in precision/sensitivity with RR, but to make bam files consistent with other projects it is on here. Actually, realignment helps samtools to call indels better.
  * no bed file. Let callers call every variant which has coverage, we will filter poorly covered variants later. Modern exome capture kits are so perfect, that
we can discover a useful non-coding variant. No sense to filter them out during that stage.
  * effects: VEP. There is a holywar VEP/snpEff/Annovar. My choice is VEP. You will see later both Ensembl and Refseq in the report, so no reason for using Annovar.
  * effects_transcripts: all. We want all effects of a variant on all transcripts to be reported.
  * aligner: bwa. Even staring with bam files, bwa is used. Sometimes input bam files aligned against older reference, or different (chr) naming scheme. It is better to have a bam file consistent with calls made.
  * custom annotation [cre.vcfanno.conf](../master/cre.vcfanno.conf) using data sources installed in bcbio.
  * creates gemini database with vcf2db

## 3a. Input files are in Illumina basespace.
* use basespace-cli to dump bcl files to HPC, then do 1b.

## 3b. Input is Illumina run (bcl files).
* create a sample sheet and run [bcl2fq.sh](../master/bcl2fq.sh).

## 3c. Input is cram file.
* Run [cram2fq.sh](../master/cram2.fq). 
* I would suggest to avoid crams when possible. A damaged bam file could be recovered with [cre.bam_recovery.sh](../master/cre.bam_recovery.sh), but nothing could be done for cram.

# 4. Run bcbio

* Single project: `qsub ~/cre/bcbio.pbs -v project=[project_name]`
Project should have a folder project_name in the current directory.

* Multiple projects: `qsub -t 1-N ~/cre/bcbio.array.pbs`
Current directory should have a list of projects in projects.txt.

# 5. Clean result dir and create project.csv report: 

`qsub ~/cre/cre.sh -v family=[family],cleanup=1`
 * moves project results and sample bam files to family dir
 * removes work and final dirs from bcbio project
 * removes gemini databases for individual callers (we need only ensemble gemini database)

  During the report generation step:
  * dumps variants from gemini database to tab text file
  * dumps variant impacts from gemini database to tab text file
  * annotates variants with refseq in addition to ensembl
  * gets coverage from GATK Haplotype calls, freebayes, and platypus
  * build excel report based on gemini variants table, variant impacts, coverage information and some other fields.

# 6. Step 5 in detail

6.1 [Report description](https://docs.google.com/document/d/1zL4QoINtkUd15a0AK4WzxXoTWp2MRcuQ9l_P9-xSlS4/edit?usp=sharing).\
6.2 [Report example for Ashkenazim trio from NIST](https://drive.google.com/open?id=0B_bLL10GwDnsN29vY3RRdGlXMWM). \
6.3 [gemini.gemini2txt.sh](../master/gemini.gemini2txt.sh) [project-ensembl.db] - dumps a gemini database into text file. \
6.4 [gemini.variant_impacts.sh](../master/gemini.variant_impacts.sh) [project-ensembl.db] dumps variant impacts from gemini. \
6.5 creates a vcf file with rare and potentially deleterious variants, the same set of variants is in the excel report.
```
cat ${family}-ensemble.db.txt | cut -f 23,24  | sed 1d | sed s/chr// > ${family}-ensemble.db.txt.positions
    bcftools view -R ${family}-ensemble.db.txt.positions -o ${family}.vcf.gz -O z ${family}-ensemble-annotated-decomposed.vcf.gz

```
6.6 coverage from VCFs produced by GATK, platypus, and freebayes - requires gatk wrapper from bcbio.
```
vcf.freebayes.getAO.sh ${family}-freebayes-annotated-decomposed.vcf.gz
vcf.gatk.get_depth.sh ${family}-gatk-haplotype-annotated-decomposed.vcf.gz
vcf.platypus.getNV.sh ${family}-platypus-annotated-decomposed.vcf.gz

```
6.7 Rscript ~/cre/[cre.R](../master/cre.R) [family] - creates report family.csv.

# 7. How to create a database of variants

- cre.database.sh [input_dir] [output_dir] - creates sample-wise and variant-wise reports, which are necessary for annotation with cre.R.
- cre.database.pull_gene.sh [database_prefix] [gene_name] - pull a gene report from the database.

# 8. Coverage plots

- ~/bioscripts/genes.R - pull a bed file from Ensembl
- ~/bioscripts/bam.coverage.bamstats05.sh - calculate coverage
- cheo.R - plot coverage pictures

# 9. List of all scripts

* bcbio.array.pbs
* bcbio.pbs
* bcbio.prepare_families.sh
* bcbio.rename_old_names.sh
* bcl2fastq.sh
* cheo.R - mostly Venn diagrams to compare pipeline validations + some coverage analysis
* cre.prepare_bcbio_run.sh
* cre.bam.validate.sh
* cre.bam.remove_decoy_reads.sh - removes decoy reads for grch37d5 with VariantBam and grep.
* cre.bcbio.upgrade.sh - examples of bcbio installation and upgrade
* cre.coverage.bamstats05.sh - calculate coverage
* cre.fixit.sh - fixes sample names
* cre.gemini_load.sh loads vep-annotated vcf to gemini db.
* cre.gemini.get_variants4gene.sh - pull all varaints for a specific gene.
* cre.gnomad_scores.R - download and parse gnomad scores.
* cre.immunopanels.R - annotates CRE report with 6 immunopanels.
* cre.kinship.R - to plot relatedness (kinship) diagram for a group of samples. Sometimes helps to detect and solve mislabelling.
* cre.package.sh
* cre.rtg.validate.sh - validates NA12878 calls vs genome in a bottle callset with RTG and a bed file
* cre.sh - master script to produce variant reports from bcbio output
* cre.topmed.R - pull variant frequency from TopMed having rs_id
* cre.roh.h3m2.sh: a robust method of ROH/LOH analysis with h3m2, calls variants, accounts for exonic regions, LD, plots picture.
* cre.roh.naive.sh: retrieves MAF<5% high quality variants from gemini.db and reports stretches of >9 HOM variants (start, end, length, genes).
* cre.vcf.has2dp.sh fixes input vcf file from HAS pipeline (Illumina, TCAG) filling DP field
* omim.sh
* orphanet.sh
* vcf.freebayes.getAO.sh
* vcf.gatk.get_depth.sh
* vcf.platypus.getNV.sh
* vcf.samtools.get_depth.sh
* vcf.split_multi.sh
* vep4seqr_hg38.sh
* vep4seqr.sh

# 10. Credits

This work was inspired by 
* [bcbio](https://github.com/chapmanb/bcbio-nextgen/) and [gemini](https://github.com/arq5x/gemini) teams. Thank you all!
* Kristin Kernohan from Children's Hospital of Eastern Ontario (CHEO), who generated most ideas about the report contents. Thank you, Kristin, for all of the discussions!

Thank you colleagues at [CCM](https://ccm.sickkids.ca/), for seminars and personal discussions.


HISTORY.md
## 0.0.4 (23 May 2019)
- fixed a bug in clinvar column

## 0.0.3 (21 May 2019)
- bug fixes, WES report generated for NA12878 looks ok

## 0.0.1 (15 May 2019)
- initial release, compatible with bcbio_1.1.5 and gnomad2.1

## previous notes
- 2018-11-07: new gnomad obseverved/expected scores instead of pLi and exac_missense. cre.gnomad_scores.R
- 2018-11-02: added back Gerp_score, updated OMIM
- 2017-09-22: added Info_refseq and Maf_exac to the database report
- 2017-09-14: improved cre.database.sh: it creates databases for cre.R
- 2017-07-13: added cre.package.sh. It packages reports to send.
- 2017-07-07: added bam.validate.sh, it produces bam.check file with picard's validation status.
- 2017-07-04: bcbio uses vep-merged, a merged cache of refseq and vep, no need to annotate with refseq anymore, removed refseq from cre.sh.


cheo.R
#https://cran.r-project.org/web/packages/VennDiagram/VennDiagram.pdf
pipeline_comparisons <- function(){
  
    library("VennDiagram")
    venn.plot <- draw.pairwise.venn(945+4958,132+4958,4958,
                c("BCBIO","Jacek"), fill=c("blue","red"), lty="blank",
              cex = 2, cat.cex=2, cat.just = list(c(-1,-1),c(1,1)),
              ext.length = 0.3,  ext.line.lwd=2,
              ext.text = F, main="Sdf"
              )
    venn.plot <- draw.pairwise.venn(46807953, 54328695, 44991414,
                             c("Nimblegen.capture", "Agilent"),
                             fill = c("blue","red"), cex = 2, cat.cex = 2)

    venn.plot3 <- draw.triple.venn(54328695, 46807953, 65824553,
                            39517373, 43777891, 44991414,
                            37679751,
                            c("Agilent", "Nimblegen.capture", "Nimblegen.empirical")
    )

    venn.plot3 <- draw.triple.venn(1, 2, 3, 12, 23, 13, 123,
                                   c("Agilent", "Nimblegen.capture", "Nimblegen.empirical")
    )

    #example
    for (pipeline in c("genap", "bcbio", "jacek", "cpipe")){
        for (sample in c("S1", "S2", "S4", "S5", "S6", "S7", "S8")){
            agilent <- unlist(read.table(paste0("S1.", pipeline, ".agilent.omim.variants.indels")))
            nimblegen <- unlist(read.table(paste0("S1.", pipeline, ".nimblegen.omim.variants.indels")))
            #illumina=unlist(read.table(paste0("S1.",pipeline,".illumina.txt")))

            x <- list(agilent, nimblegen)
            names <- list("Agilent", "Nimblegen")
            png(paste0(sample, ".", pipeline, ".png"))
            grid.draw(venn.diagram(x, NULL, category.names = names))
            dev.off()
        }
    }

    for (platform in c("agilent", "nimblegen")){
        for (sample in c("S1", "S2", "S4", "S5", "S6", "S7", "S8")){
            #sample="S3"
            #platform="illumina"
            jacek <- unlist(read.table(paste0(sample, ".jacek.", platform, ".omim.variants.indels")))
            cpipe <- unlist(read.table(paste0(sample, ".cpipe.", platform, ".omim.variants.indels")))
            bcbio <- unlist(read.table(paste0(sample, ".bcbio.", platform, ".omim.variants.indels")))
            genap <- unlist(read.table(paste0(sample, ".genap.", platform, ".omim.variants.indels")))

            x <- list(jacek, cpipe, bcbio, genap)
            names <- list("jacek", "cpipe", "bcbio", "genap")

            png(paste0(sample, ".", platform, ".png"))
            grid.draw(venn.diagram(x, NULL, category.names = names))
            dev.off()
        }
    }

}

intersect_bed_files <- function(){
    # http://davetang.org/muse/2013/01/02/iranges-and-genomicranges/
    source("http://bioconductor.org/biocLite.R")
    biocLite("GenomicRanges")
    library("GenomicRanges")
    setwd("coverage/nimblegen")
    capture <- read.table("nimblegen.capture", header = F)
    colnames(capture) <- c("chr", "start", "end")
    capture.bed <- with(capture, GRanges(chr, IRanges(start+1, end)))

    empirical <- read.table("nimblegen.empirical", header = F)
    colnames(empirical) <- c("chr", "start", "end")
    empirical.bed <- with(capture, GRanges(chr, IRanges(start+1, end)))

    omim <- read.table("omim.orphanet.goodnames.v2.bed", header = F)
    colnames(empirical) <- c("chr", "start", "end")
    omim.bed <- with(capture, GRanges(chr, IRanges(start+1, end)))

    bed.intersect <- intersect(omim.bed, capture.bed)

    #bed files
    setwd("venn_diagrams/bed_intersection/")
    omim <- unlist(read.table("omim.orphanet.goodnames.v2.bed.exons"))
    agilent <- unlist(read.table("omim_vs_agilent.50percent.wo.exons"))
    nimblegen <- unlist(read.table("omim_vs_nimblegen.capture.50percent.wo.exons"))
    illumina <- unlist(read.table("omim_vs_illumina.50percent.wo.exons"))

    x <- list(omim,agilent,nimblegen,illumina)
    names <- list("omim","agilent","nimblegen","illumina")

    grid.draw(venn.diagram(x, NULL, category.names = names))
}


variants_parameter <- function(){
    type <- "snps"
    type <- "indels"
    for (platform in c("agilent", "nimblegen")){
        for (sample in c("S1", "S2", "S4", "S5", "S6", "S7", "S8")){
            sample <- "S1"
            platform <- "agilent"
            cpipe.hom <- read.table(paste0(sample, ".cpipe.", platform, ".omim.", type, ".hom.AD"))
            cpipe.het <- read.table(paste0(sample, ".cpipe.", platform, ".omim.", type, ".het.AD"))
    
            bcbio.hom <- read.table(paste0(sample, ".bcbio.", platform, ".omim.", type, ".hom.AD"))
            bcbio.het <- read.table(paste0(sample, ".bcbio.", platform, ".omim.", type, ".het.AD"))
    
            genap.hom <- read.table(paste0(sample, ".genap.", platform, ".omim.", type, ".hom.AD"))
            genap.het <- read.table(paste0(sample, ".genap.", platform, ".omim.", type, ".het.AD"))

            genap.only <- read.table("S1.genap.only.recode.vcf.AD")
            v <- c(genap.hom, genap.het, genap.only)
            names <- c("genap.hom", "genap.het", "genap.only")
    
        #indels
        v <- c(cpipe.het, bcbio.het, genap.het)
        names <- c("cpipe.het", "bcbio.het", "genap.het")
    
        png(paste0(sample,".", platform, ".", type, ".png"), width = 1000)
        boxplot(v, names = names)
        dev.off()
      }
    }

    setwd("~/cluster/dorin_test")
    b1 <- read.table("UNIQUE_to_BCBIO_1496461.vcf.AD")
    b2 <- read.table("UNIQUE_to_BCBIO_1496462.vcf.AD")
    b3 <- read.table("UNIQUE_to_BCBIO_1496463.vcf.AD")
    g1 <- read.table("UNIQUE_to_DNASEQ_1496461.vcf.AD")
    g2 <- read.table("UNIQUE_to_DNASEQ_1496462.vcf.AD")
    g3 <- read.table("UNIQUE_to_DNASEQ_1496463.vcf.AD")

    v <- c(b1, b2, b3, g1, g2, g3)
    names <- c("b1", "b2", "b3", "g1", "g2", "g3")

    png(paste0(sample, ".", platform, ".", type, ".png"), width = 1000)
    boxplot(v, names = names)
}

# title = "cheo.omim_genes.coverage"
# coverage.gene_panel(title)
# plots coverage for every gene for all samples in samples.txt, each sample should have sample.coverage - output of
# bam.coverage.bamstats05.sh
coverage.gene_panel <- function(title){
    #test
    title <- "test"
    setwd("~/Desktop/work")
    files <- list.files(".", "\\.coverage$")
    #samples = unlist(read.table("samples.txt", stringsAsFactors=F))
    
    coverage <- read.delim(files[1], header = T, stringsAsFactors = F)
    coverage <- coverage[,c("gene", "avg")]
    colnames(coverage)[2] <- files[1]

    for (file in tail(files,-1)){
        sample_coverage <- read.delim(file,header=T,stringsAsFactors = F)
        sample_coverage <- sample_coverage[,c("gene", "avg")]
        colnames(sample_coverage)[2] <- file
        coverage <- cbind(coverage,sample_coverage[2])
    }
    row.names(coverage) <- coverage$gene
    coverage$gene <- NULL
    
    n_genes <- nrow(coverage)
    
    for(i in seq(1, ceiling(n_genes/100))){
         start_index <- (i-1)*100+1
         end_index <- i*100
         
         if (end_index > n_genes) end_index <-  n_genes
         
         png(paste0(title, ".part", i, ".png"), res = 300, width = 5000, height = 2000)
         boxplot(t(coverage[start_index:end_index,]), las = 2, cex.axis = 0.8,
                 main = paste0("Coverage in ", length(files), " samples of NextSeq for ", 
                               title, " gene panel,part ", i))
         dev.off()
    }
}

#when looking at all genes, some samples may have no coverage
# CA0229.coverage 15325
# CA0246.coverage 17268
# CH0317.coverage 15833
# GM15262.coverage 14435
coverage.all_genes <- function (){
    title <- "Coverage in project 913 across all protein coding genes,no outliers"
    library("matrixStats")
    setwd("~/Desktop/work")
    samples <- unlist(read.table("samples.txt", stringsAsFactors = F))
  
    #hopefully 1st sample has most genes
    coverage <- read.delim(paste0(samples[1], ".coverage"), header = T, stringsAsFactors = F)
    coverage <- coverage[,c("gene","mean")]
    colnames(coverage)[2] <- samples[1]
    row.names(coverage) <- coverage$gene
    coverage$gene <- NULL
  
    for (sample in tail(samples,-1)){
        sample_coverage <- read.delim(paste0(sample,".coverage"),header=T,stringsAsFactors = F)
        sample_coverage <- sample_coverage[,c("gene","mean")]
        colnames(sample_coverage)[2] <- sample
        coverage <- merge(coverage, sample_coverage, by.x = "row.names", by.y = "gene", all.x = T)
        row.names(coverage) <- coverage$Row.names
        coverage$Row.names <- NULL
    }
  
    coverage[is.na(coverage)] <- 0  
    coverage$Mean <- rowMeans(coverage)
    png("coverage.png", res = 300, width = 5000, height = 2000)
    boxplot(coverage, las = 1, cex.axis = 0.6,
            main = title, outline = F)
    dev.off()
    
    meds <- rbind(colnames(coverage), colMedians(as.matrix(coverage)))
    write.table(meds, "medians.txt", col.names = F, quote = F, row.names = F)
}

#%of bases covered more than 10x
coverage.percent_more_than10x <- function(){
    setwd("~/Desktop/work")
    samples <- unlist(read.table("samples.txt", stringsAsFactors = F))
  
    for (sample in samples){
        coverage <- read.delim(paste0(sample, ".coverage"), header = T, stringsAsFactors = F)
        total_len <- sum(coverage$length)
    
        coverage_10x <- coverage[coverage$mean > 10,]
        len_10x <- sum(coverage_10x$length)
    
        print(paste0(sample, " ", len_10x/total_len))
    }
}

omim_table_manipulation <- function(){
    setwd("~/Desktop/reference_tables/OMIM_2017-04-13/")
    mimTitles.percent <- read.delim2("mimTitles.percent.txt", comment.char="#")    
    genemap2 <- read.delim("genemap2.txt", comment.char="#")
    mimTitles.percent <- merge(mimTitles.percent, genemap2, by.x = "Mim.Number", 
                               by.y = "Mim.Number", all.x = T, all.y = F)
    write.csv(mimTitles.percent, "mimTitles.percent.location.csv", row.names = F)
}

read_length_distribution <- function(family){
    
    family_data <- read.delim(paste0(family,".tsv"), stringsAsFactors = F)
    samples <- unique(family_data$sample)
    
    for (sample in samples){
        tmp <- subset(family_data, sample == sample)
        tmp$sample <- NULL
        print(paste0(sample, " ", round(sum(tmp$Length*tmp$Count) / sum(tmp$Count))), quote = F)
    }
}

read_lengths <- function(){
    setwd("~/Desktop/project_cheo/2017-04-12_read_lengths/")
    families <- unlist(read.table("projects.txt", stringsAsFactors = F))
    for (family in families){
      read_length_distribution(family)
    }

    read_lengths <- read.csv("read_lengths.txt", sep="", stringsAsFactors = F)
    read_lengths$id <- NULL

    png("read_lengths_all_samples.png", width = 2000)
    barplot(read_lengths$average_read_length, names.arg = read_lengths$sample,
            main = "Average read lengths for NextSeq samples is 134",
            las = 2)
    dev.off()
}
